### DISCUSSION TITLE
Transaction

### DISCUSSION THREAD

**Discussion by panarch** [gluesql/gluesql#228] [Ideas]
This kind of pseudo code below may work I think.
Introducing two transaction traits; `BeginTransaction` and `EndTransaction`.
```rust
trait BeginTransaction where T: Store + StoreMut + EndTransaction {
  fn begin_transaction -> MutResult<T, ()>;
}

trait EndTransaction where T: Store + StoreMut + BeginTransaction {
  fn rollback_transaction -> MutResult<T, ()>;
 
  fn commit_transaction -> MutResult<T, ()>;
}
```

Unlike other traits like `StoreMut` or `AlterTable`, `BeginTransaction` trait returns `Store + StoreMut + EndTransaction` rather than `Self`.
Then, in the case of `SledStorage`.. (I'll skip including the optional ones `Index` and `AlterTable`)
We can make `SledStorage` to implement `Store`, `StoreMut` and `BeginTransaction`.
Note that using existing `StoreMut` implementation cannot store the intermediate data for transaction rollback situation.
To make this work, we can add `TransactionalSledStorage` which implements `Store`, `StoreMut` and `EndTransaction`.

`SledStorage` can only begin transaction.
`TransactionalSledStorage` can only rollback or commit transaction.
Then, we can make `TransactionalSledStorage` stateful itself by storing the data during transaction.

**Comment by rnbguy** [comment]
Going through Sled's transactional implementation, it seems your `TransactionalSledStorage` is analogous to `TransactionalTree` of Sled.

https://docs.rs/sled/0.34.6/sled/transaction/struct.TransactionalTree.html

I will look into it :) thanks

**Comment by KyGost** [comment]
Hmmm I think this would be a nice easy work around for the current structure but inferior in the end.

While it would be quite easy to implement, especially for something like Sled, I think it would be better to have operate transactions on the Glue side, if I remember to I will elaborate further tomorrow.
